### ISSUE TITLE
pyserial on raspberry pi with socat raises IOError on open

### ISSUE THREAD

**Issue by acencini** [pyserial/pyserial#59]
using socat on Raspberry Pi (wheezy), Python 2.7, pyserial 3.0, doing an open on a socat serial port raises the following exception:
```
serial.Serial(port="/dev/ttyVapor002")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/serial/serialutil.py", line 171, in __init__
    self.open()
  File "/usr/local/lib/python2.7/dist-packages/serial/serialposix.py", line 311, in open
    self._update_dtr_state()
  File "/usr/local/lib/python2.7/dist-packages/serial/serialposix.py", line 605, in _update_dtr_state
    fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_DTR_str)
IOError: [Errno 22] Invalid argument
```

/dev/ttyVapor002 is hooked up to /dev/ttyVapor001 for a serial emulator setup using socat.  We were using pyserial 2.7 previously and quite happily without error; on upgrade to pyserial 3.0, it appears the switch in how DTR state is updated does not work on this platform.  Happy to dig around to help find a fix - it doesn't appear DTR state is set during open in pyserial-legacy, though setDTR does exist in serialposix.py there.

**Comment by acencini** [comment]
we have confirmed this is an issue on linux x86/64 with socat as well

**Comment by zsquareplusc** [comment]
Can you give an example command line for `socat` in your use case? That tool has an awful lot of options ;-)

pySerial 3.0 now applies the RTS/DTR control state on open, unless hardware flow control is used (which may be a workaround for you, to set that). I guess it would have failed with older pySerial versions too if the control line was toggled, e.g. by the application?

**Comment by acencini** [comment]
thanks for the response - the command line is as follows:

`socat PTY,link=/dev/ttyVapor001,mode=666 PTY,link=/dev/ttyVapor002,mode=666`

the suggestion to use hardware flow control actually fixed the issue - thank you!  in pyserial 3.0, i was able to open the socat device by setting `rtscts=True` and `dsrdtr=True` which succeeded, whilst without those settings, it failed.

**Comment by austinphillips2** [comment]
@zsquareplusc 

You indicated:

> pySerial 3.0 now applies the RTS/DTR control state on open, unless hardware flow control is used

However in serialposix.py, the following occurs on open:
```python
        if not self._dsrdtr:
            self._update_dtr_state()
        if not self._rtscts:
            self._update_rts_state()
```

Should the logic of these conditionals be inverted, or am I reading this incorrectly?

For example, `self._rtscts == True` should indicate that RTS/CTS hardware flow control is enabled, however in the current implementation RTS/CTS state is only updated when `_rtscts` is False.

**Comment by zsquareplusc** [comment]
You read that correctly. In the case of hardware flow control, the hardware takes over the control :-) so modifying the states by software does not change the state of the pin. Only when hardware flow control is disabled, it makes sense to control the state by software.

**Comment by austinphillips2** [comment]
Thanks for the clarification.

It seems to me that the flow control for RTS/CTS should be a 3 state instead of the current 2 states.

 1. None.  No hardware flow control, either software controlled, or UART controlled. (This is currently not supported)
 2. Hardware controlled.  UART hardware automatically manages flow control lines.  Currently `Serial(port, rtscts=True)`
 3. Software controlled.  Flow control lines are controlled via software commands. Currently `Serial(port, rtscts=False)`.  This doesn't seem intuitive as flow control is enabled even though r`tscts=False`

Allowing for option 1 above would allow the use of the the `Serial` class on tty based virtual ports as others have attempted in this thread which aren't backed by a full serial driver with `ioctl` support.  As it stands, there is no way to avoid the device specific `ioctl` calls unless `Serial(port, rtscts=True)` is used.

Should there be some kind of flag which can be passed to the constructor of `Serial` which can be used to specify the three types of flow control?

eg. To configure the 3 modes as above.

 1. `Serial(port, rtscts=False)`.  No hardware flow control.
 2. `Serial(port, rtscts=True)` or `Serial(port, rtscts=True, flags = RTSCTS_HARDWARE)`.  Would default to use of hardware based flow control if no flags specified.
 3. `Serial(port, rtscts=True, flags = RTSCTS_SOFTWARE)`.  Use hardware flow control managed via software rather than via UART.

This would allow for intuitive use of two-wire interfaces that don't have any hardware flow control (eg com ports emulated through use of ttys) and full serial interfaces.

**Comment by zsquareplusc** [comment]
I have checked in a possible fix. On posix, the control lines are only set if `.rts`/`.dtr` are set before `open()`. Reading the attributes before first setting them will return `None`.

This fix should solve that setting the control lines before opening is supported (new in 3.x) for those that need it and not setting anything otherwise (old 2.x behavior).

Unfortunately this may cause that on posix the control lines are not activated when the port is opened, while on Windows it does that (with HW flow control disabled in both cases) :/

**Comment by austinphillips2** [comment]
Although I haven't tried your fix, this seems like a reasonable solution even though it doesn't quite match behaviour in Windows.

For those that require software controlled rts/dtr flow control to operate the same on all platforms, they could achieve this by using `Serial()` without specifying a port (so that an implicit open doesn't occur on __init__), then set rts/dtr appropriately followed by an explicit open.

**Comment by zsquareplusc** [comment]
Yes Posix and Windows (and `serialcli`) behave differently in respect to the control lines on open, but so they did in pySerial V2.x. The attempt to align them in one way in V3.x has caused the problems above...

Btw., my guess is that almost no one does "software" flow control with the control lines. Most users need them to communicate with embedded devices / micro-controllers where the control lines are used to activate boot loaders or change the device behavior.

**Comment by davidjb** [comment]
Upgrading to the latest `master` version of pyserial resolves this issue, so it seems the fix is working for me (Raspberry Pi, USB ZTE K3571 modem, Python 2.7); thanks!

**Comment by Peque** [comment]
I created a very simple test case to prevent this happening again in the future (#111). Any comments/suggestions on how to improve that pull request (if necessary) would be more than welcome. :blush:

**Comment by zsquareplusc** [comment]
released 3.1 where an error setting control lines in open() is ignored (but not in later calls).

**Comment by akilawickey** [comment]
Thanks i fixed my issue using 

by setting rtscts=True and dsrdtr=True
